### PR TITLE
introduces -Yignore-crashes

### PR DESCRIPTION
**Pull request by xeno-by**
Quite useful when prototyping significant changes in the compiler
and not wanting to expose oneself to huge crash messages.

### CODE REVIEW

**Comment by xeno-by** [comment]
review @retronym

File: src/compiler/scala/tools/nsc/Global.scala [hunk 1]
```diff
@@ -1553,8 +1553,14 @@ class Global(var currentSettings: Settings, var reporter: Reporter)
...
         } else ex.getClass.getName
         // ex.printStackTrace(Console.out) // DEBUG for fsc, note that error stacktraces do not print in fsc
-        globalError(supplementErrorMessage("uncaught exception during compilation: " + shown))
-        throw ex
+        if (!settings.ignoreCrashes.value) {
+          globalError(supplementErrorMessage("uncaught exception during compilation: " + shown))
+          throw ex
+        } else {
+          var msg = if (ex.getMessage != null) ex.getMessage else ex.getClass.toString
+          if (msg.contains(scala.compat.Platform.EOL)) msg = msg.substring(0, msg.indexOf(scala.compat.Platform.EOL))
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/Global.scala]:
``` scala
scala> "ab\ncd".lines.take(1).mkString
res0: String = ab
```

**Comment by retronym** [comment]
I'm not sure this change is generally useful enough to warrant inclusion and a new `-Y`.

This section of code does have a pre-existing problem I've been meaning to fix: sometimes `supplementErrorMessage` gets called twice as an exception bubbles up.

File: src/compiler/scala/tools/nsc/Global.scala (hunk 1 above)
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/Global.scala]: Thanks!

**Comment by xeno-by** [comment]
PLS REBUILD ALL

**Comment by xeno-by** [comment]
@retronym I'd say this pull request has merit, because otherwise debugging a crashing compiler becomes really annoying. You print a line or two of debug output, then it crashes, producing a 40-60 lines log, and you're forced to scroll up to see what's going on.

If we agree that this is a worthy problem to solve, then a possible workaround would be `scalac ... 2>/dev/null`. However I wouldn't say that it's an adequate replacement, because it swallows all errors, not only fatal crashes.

**Comment by retronym** [comment]
But it's the sort of change you can make on your branch while you're working, and then scrub from the history before you submit the branch.

**Comment by xeno-by** [comment]
Well, sure, that's what I did some time ago. But when I had to do exactly the same when working on type macros, I started to feel that the situation can be improved.

**Comment by paulp** [comment]
I'm very much against adding any more options to control debugging output. We should remove about ten of them instead. In this specific case, I'm even more against it, because the output needs fixing for everyone, not just the two people who will use the obscure -Y option. "Last tree to typer" should limit its output to maybe 10 lines, and then there is the multiple-printing issue exacerbating everything.

**Comment by xeno-by** [comment]
All right, guys, next time I'll come up with something more substantial.

**Comment by paulp** [comment]
If you're going to be around next week in lausanne, I would be up for some strategizing in this area.

**Comment by xeno-by** [comment]
Yeah I'm in Lausanne, usually at my office on weekdays from 10-11 to 18-20. Would be great to chat!

On what day of week do you arrive? I'm not sure whether I can be at EPFL next Monday, but if you're going to be there I'll figure something out.

**Comment by paulp** [comment]
Arriving tuesday afternoon. I leave saturday morning, but I'll most likely have time on friday at least.

**Pull request closed without merging**